### PR TITLE
Play console actions back in regular order

### PR DESCRIPTION
**Pull request by gtritchie**
Previously when refreshing UI, we played the ConsoleActions back in reverse order, until our maximum number of console lines was displayed. Server keeps up to 1000 console actions, but each console action can generate multiple lines of output, this is why it was done this way.

This reverse playback made client Ansi parsing more difficult, often breaking reload of console with Ansi color output, and the resulting DOM had many more nodes than the DOM created during regular usage of the console, impacting performance. Another consequence was any special parsing done on console DOM had to account for both possible structures (regular and reloaded). There are some bugs in that area, and wanted to do this change before getting to those.

Now we do an initial pass through the ConsoleActions in reverse order and count newlines to determine line count, but don't output anything to the DOM. Then do a forward pass starting where it will stop at maximum desired console length (plus potential slight overrage due to size of final action's output). That is written out in normal order, thus using the same output logic as regular console output. Finally trim any excess when the reload is done.

Anecdotally, the reduced simplicity of the resulting DOM should easily overcomes the cost of doing the initial counting pass, and in general I expect this to help with reloading large console buffers.

This does fix the Ansi code parsing issue on reload.

### CODE REVIEW

File: src/gwt/src/org/rstudio/studio/client/common/shell/ShellWidget.java [hunk 1]
```diff
@@ -457,24 +454,60 @@ private boolean trimExcess()
...
+      for (; revIndex >= 0; revIndex--)
+      {
+         ConsoleAction action = actions.get(revIndex);
+         
+         if (action.getType() == ConsoleAction.INPUT)
+            lines++;
+         
+         lines = lines + StringUtil.newlineCount(action.getData());
+         
+         excessLines = excessLines(lines);
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/common/shell/ShellWidget.java]: nit: could we just omit the excessLines variable and function altogether and just say `if (lines > maxLines_)` for simplicity?

File: src/gwt/src/org/rstudio/studio/client/common/shell/ShellWidget.java [hunk 2]
```diff
@@ -499,12 +532,21 @@ public boolean execute()
                      break;
                }
                if (!canContinue)
-                  return false;
+               {
+                  break;
+               }                  
             }
-            if (!DomUtils.selectionExists())
-               scrollPanel_.scrollToBottom();
```
Comment (jmcphers) [inline src/gwt/src/org/rstudio/studio/client/common/shell/ShellWidget.java]: When I profiled the console replay performance a while back, this turned out to be one of the biggest contributors to the replay time, so I am delighted to see it gone! Now that we play everything back in order, does the scroll panel keep up automatically (or is some other mechanism taking care of scrolling to the bottom once replay is done?)

**Comment by gtritchie** [comment]
The "excessLines" variable was left over from a slightly different approach, so yes, I'll get rid of it with my next PR.

Yes, the resizeCommand_.nudge() at the end of ShellWidget::output ensures the pane is scrolled to the bottom after reloading.

**Comment by jmcphers** [comment]
Great, thanks!

**Pull request merged**